android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIMaskedImage.java

```diff
@@ -0,0 +1,1399 @@
+			if (this.tintShader != null) {
+				Matrix matrix = new Matrix();
+				matrix.setRectToRect(new RectF(), drawableBounds, Matrix.ScaleToFit.FILL);
+				this.tintShader.setLocalMatrix(matrix);
+			}
+		}
+
+		@Override
+		public void onBlendModeChanged()
+		{
+			getPaint().setShader(null);
+		}
+
+		@Override
+		public void onTintEnabledChanged()
+		{
+			this.tintShader = null;
+		}
+
+		@Override
+		public void onTintColorChanged()
+		{
+			this.tintShader = null;
+		}
+
+		@Override
+		public void onMaskDrawableChanged()
+		{
+			this.maskShader = null;
+		}
+
+		@Override
+		public boolean canDraw()
+		{
+			return canDraw(getMaskedDrawable());
+		}
+
+		public static boolean canDraw(MaskedDrawable maskedDrawable)
+		{
+			if (maskedDrawable == null) {
+				return false;
+			}
+			if (maskedDrawable.isTintingEnabled() == false) {
+				return false;
+			}
+			if ((maskedDrawable.getMaskDrawable() instanceof BitmapDrawable) == false) {
+				return false;
+			}
+			return true;
+		}
+	}
+
+	/**
+	 * Handler optimized in drawing an image to a mask via a GPU shader. Only supports modes SRC_IN and DST_IN.
+	 * <p>
+	 * This class is only intended to be used by a "MaskedDrawable" type.
+	 */
+	private static class FastImageMaskHandler extends BaseMaskHandler
+	{
+		/** Shader used to fill the mask with an image. */
+		private BitmapShader bitmapShader;
+
+		/** The mask stored in 8-bit grayscale form. */
+		private Bitmap grayscaledBitmap;
+
+		public FastImageMaskHandler(MaskedDrawable maskedDrawable)
+		{
+			super(maskedDrawable);
+
+			// This flag prevents stretched images from being pixelated when calling Canvas.drawBitmap().
+			Paint paint = getPaint();
+			paint.setFlags(paint.getFlags() | Paint.FILTER_BITMAP_FLAG);
+		}
+
+		@Override
+		public boolean drawTo(Canvas canvas)
+		{
+			// Validate.
+			if (canvas == null) {
+				return false;
+			}
+			if (canDraw() == false) {
+				return false;
+			}
+
+			// Fetch the assigned blend mode.
+			PorterDuff.Mode blendMode = getMaskedDrawable().getBlendMode();
+
+			// Fetch/Create the 8-bit grayscaled bitmap mask.
+			if (this.grayscaledBitmap == null) {
+				Bitmap bitmap = null;
+				if (blendMode == PorterDuff.Mode.SRC_IN) {
+					bitmap = getBitmapFrom(getMaskedDrawable().getMaskDrawable());
+				} else if (blendMode == PorterDuff.Mode.DST_IN) {
+					bitmap = getBitmapFrom(getMaskedDrawable().getImageDrawable());
+				}
+				if (bitmap != null) {
+					try {
+						// If mask is not an 8-bit grayscale bitmap, then convert it into one.
+						if (bitmap.getConfig() != Bitmap.Config.ALPHA_8) {
+							bitmap = bitmap.extractAlpha();
+						}
+
+						// Android 4.4 has a bug where if bitmap byte width does not fit a 4 byte packing alignment,
+						// then mask will appear skewed. Resizing bitmap doesn't work. So, give up if this happens.
+						if ((Build.VERSION.SDK_INT == 19) || (Build.VERSION.SDK_INT == 20)) {
+							if ((bitmap.getWidth() % 4) != 0) {
```

Comment:
Could this be moved above the `bitmap.extractAlpha()` to fail before reaching `bitmap.extractAlpha()` ?

Reply:
Ahh, good catch. I was trying to "fix" the resulting grayscaled bitmap, but later gave up.